<div align="center"> 

**hopa1029, klfl1012**

<img src="s_purple.png" alt="drawing" width="2000" height= 200/>


<div align="left">

# Übungsblatt 2 - Personalabteilung
__________________________________________________________________________________________________________________________________________________________________________________________________________________

**DSCB310 – Datenanalyse und Business Intelligence 1**  

Wintersemester 22/23

##### Gliederung des Notebooks:

1. Ziel der Datenanaylse für die Personalabteilung
   
2. Fragestellungen:

   - P1: Prüfen Sie, ob ein Zusammenhang zwischen dem Trinkgeldgeben und dem Vorhandensein eines der folgenden Produkte in einer Bestellung existiert: 24852, 24964, 2120
  
   - P2: Lassen sich regionale Unterschiede im Trinkgeldverhalten erkennen?
  
   - P3: Welche Attribute einer Bestellung wirken sich auf das Trinkgeldverhalten aus?
  
   - P4: Spielt die Vergangenheit eines Users eine Rolle in Hinblick auf die Trinkgeldwahrscheinlichkeit, oder kommt es nur auf Inhalt und Parameter der aktuellen Bestellung an?
  
3. Zusätzliche Fragestellungen

##### Imports, Datensatz und Funktionen: 

**Imports:**

In [156]:
import pandas as pd
import numpy as np

import plotly.express as px
import plotly.graph_objects as go
from plotly.subplots import make_subplots

from urllib.request import urlopen
import json
import ssl
ssl._create_default_https_context = ssl._create_unverified_context

from sklearn.model_selection import train_test_split
from sklearn.tree import DecisionTreeClassifier
from sklearn.model_selection import GridSearchCV
from sklearn.metrics import accuracy_score, recall_score, precision_score
from sklearn.ensemble import HistGradientBoostingClassifier
from sklearn.pipeline import make_pipeline

**Datensatz:**

In [157]:
df = pd.read_parquet("DSCB310 - UE2 - Shopping Carts.parquet")

In [158]:
df.reset_index(drop= True).head()

order_id  product_id  add_to_cart_order  reordered  user_id  order_number  \
0         1       13176                  6          0   112108             4   
1         1       11109                  2          1   112108             4   
2         1       10246                  3          0   112108             4   
3         1       22035                  8          1   112108             4   
4         1       49683                  4          0   112108             4   

   order_dow  order_hour_of_day  days_since_prior_order  tip  \
0          4                 10                     9.0    1   
1          4                 10                     9.0    1   
2          4                 10                     9.0    1   
3          4                 10                     9.0    1   
4          4                 10                     9.0    1   

                                    product_name  aisle_id  department_id  \
0                         Bag of Organic Bananas        24              4   
1  Organic 4% Milk Fat Whole Milk Cottage Cheese       108             16   
2                          Organic Celery Hearts        83              4   
3                    Organic Whole String Cheese        21             16   
4                                 Cucumber Kirby        83              4   

   department                 aisle county  
0     produce          fresh fruits   Inyo  
1  dairy eggs  other creams cheeses   Inyo  
2     produce      fresh vegetables   Inyo  
3  dairy eggs       packaged cheese   Inyo  
4     produce      fresh vegetables   Inyo

**Funktionen:**

In [ ]:
# Einheitliche Print Funktion zur Rückgabe von accuracy, recall und precision des verwendeten Modells

def print_evaluation(pipeline_or_model: str, X_train, X_test, y_train, y_test, y_train_pred, y_test_pred):
    accurary_train = accuracy_score(y_train, y_train_pred)
    recall_train = recall_score(y_train, y_train_pred)
    precision_train = precision_score(y_train, y_train_pred)

    accurary_test = accuracy_score(y_test, y_test_pred)
    recall_test = recall_score(y_test, y_test_pred)
    precision_test = precision_score(y_test, y_test_pred)
    
    print(
        f"{pipeline_or_model} Evaluation:\n"
        f"{'':6} {'ACC':>10} | {'REC':>14} | {'PRE':>10} | {'rows':>8} | {'columns':>8}\n"
        f"{'Train':6} {accurary_train:10.5f} | {recall_train:14.5f} | {precision_train:10.5f} | {X_train.shape[0]:8} | {X_train.shape[1]:8}\n"
        f"{'Test':6} {accurary_test:10.5f} | {recall_test:14.5f} | {precision_test:10.5f} | {X_test.shape[0]:8} | {X_test.shape[1]:8}\n"
    )

## 1. Ziel der Datenanalyse für die Personalabteilung:

- Nachvollziehbare und systematische Muster bei der Trinkgeldvergabe finden

- Wovon die Wahrscheinlichkeit, bei einer Bestellung Trinkgeld zu erhalten, abhängt

-> Damit die Chancengleichheit der Fahrer verbessert wird, indem sie entweder vom Unternehmen extra entlohnt werden oder die vielversprechenden und weniger aussichtsreichen Fahrten gleichmäßig auf die Fahrer verteilt werden können.

## 2. Fragestellungen:

### **P1:** Prüfen Sie, ob ein Zusammenhang zwischen dem Trinkgeldgeben und dem Vorhandensein eines der folgenden Produkte in einer Bestellung existiert: 24852, 24964, 2120:

In [ ]:
df_tip = df.copy()

In [ ]:
df_tip.loc[df_tip.product_id.isin([24852, 24964, 2120])].product_name.unique()

##### Bestellungen mit Organic Garlic

In [ ]:
df_tip_1 = df.copy()

In [ ]:
df_tip_1["p1_products_in_order"] = 0
df_tip_1.loc[df_tip_1.product_id.isin([24852]), "p1_products_in_order"] = 1

**Bestellungen die das Produkt Organic Garlic enthalten haben eine höhere Wahrscheinlichkeit einen Tipp zu erhalten**  

In [ ]:
pd.crosstab(df_tip_1.p1_products_in_order, df.tip, margins= True, normalize= "index")

**Testen ob Unabhängig / Abhängig:**

- H0: (Nullhypothese) Die beiden Variablen sind unabhängig.
- H1: (alternative Hypothese) Die beiden Variablen sind nicht unabhängig.

In [ ]:
pd.crosstab(df_tip_1.p1_products_in_order, df.tip, margins= True)

In [ ]:
import scipy.stats as stats

data = pd.crosstab(df_tip_1.p1_products_in_order, df.tip)

stats.chi2_contingency(data)

Da der p-Wert des Tests weniger als 0.05 beträgt wird die Nullhypothese abgelehnt, es gilt H1.

In [ ]:
pd.crosstab(df_tip_1.p1_products_in_order, df.tip, margins= True, normalize= "all")

In [ ]:
# hji = hi * hj
pd.crosstab(df_tip_1.p1_products_in_order, df.tip, margins= True, normalize= "all").iloc[1, 1:2].values[0] == 0.006693 * 	0.007841

##### Bestellungen mit Banana

In [ ]:
df_tip_2 = df.copy()

In [ ]:
df_tip_2["p1_products_in_order"] = 0
df_tip_2.loc[df_tip_2.product_id.isin([24964]), "p1_products_in_order"] = 1

**Bestellungen die das Produkt Banana enthalten haben eine höhere Wahrscheinlichkeit einen Tipp zu erhalten**  

In [ ]:
pd.crosstab(df_tip_2.p1_products_in_order, df.tip, margins= True, normalize= "index")

**Testen ob Unabhängig / Abhängig:**

- H0: (Nullhypothese) Die beiden Variablen sind unabhängig.
- H1: (alternative Hypothese) Die beiden Variablen sind nicht unabhängig.

In [ ]:
pd.crosstab(df_tip_2.p1_products_in_order, df.tip, margins= True)

In [ ]:
import scipy.stats as stats

data = pd.crosstab(df_tip_2.p1_products_in_order, df.tip)

stats.chi2_contingency(data)

Da der p-Wert des Tests weniger als 0.05 beträgt wird die Nullhypothese abgelehnt, es gilt H1.

In [ ]:
pd.crosstab(df_tip_2.p1_products_in_order, df.tip, margins= True, normalize= "all")

In [ ]:
# hji = hi * hj
pd.crosstab(df_tip_2.p1_products_in_order, df.tip, margins= True, normalize= "all").iloc[1, 1:2].values[0] == 0.447496 * 0.018123

##### Bestellungen mit Sauvignon Blanc

In [ ]:
df_tip_3 = df.copy()

In [ ]:
df_tip_3["p1_products_in_order"] = 0
df_tip_3.loc[df_tip_3.product_id.isin([2120]), "p1_products_in_order"] = 1

**Bestellungen die das Produkt Sauvignon Blanc enthalten haben eine höhere Wahrscheinlichkeit einen Tipp zu erhalten**  

In [ ]:
pd.crosstab(df_tip_3.p1_products_in_order, df.tip, margins= True, normalize= "index")

**Testen ob Unabhängig / Abhängig:**

- H0: (Nullhypothese) Die beiden Variablen sind unabhängig.
- H1: (alternative Hypothese) Die beiden Variablen sind nicht unabhängig.

In [ ]:
pd.crosstab(df_tip_3.p1_products_in_order, df.tip, margins= True)

In [ ]:
import scipy.stats as stats

data = pd.crosstab(df_tip_3.p1_products_in_order, df.tip)

stats.chi2_contingency(data)

Da der p-Wert des Tests weniger als 0.05 beträgt wird die Nullhypothese abgelehnt, es gilt H1.

In [ ]:
pd.crosstab(df_tip_3.p1_products_in_order, df.tip, margins= True, normalize= "all")

In [ ]:
# hji = hi * hj
pd.crosstab(df_tip_3.p1_products_in_order, df.tip, margins= True, normalize= "all").iloc[1, 1:2].values[0] == 0.447496 * 0.018123

##### Bestellungen mit Organic Garlic, Banana und Sauvignon Blanc

In [ ]:
df_tip["p1_products_in_order"] = 0
df_tip.loc[df_tip.product_id.isin([24852, 24964, 2120]), "p1_products_in_order"] = 1

**Bestellungen die die Produkte Organic Garlic, Banana und Sauvignon Blanc enthalten haben eine höhere Wahrscheinlichkeit einen Tipp zu erhalten**  

In [ ]:
pd.crosstab(df_tip.p1_products_in_order, df.tip, margins= True, normalize= "index")

**Testen ob Unabhängig / Abhängig:**

- H0: (Nullhypothese) Die beiden Variablen sind unabhängig.
- H1: (alternative Hypothese) Die beiden Variablen sind nicht unabhängig.

In [ ]:
pd.crosstab(df_tip.p1_products_in_order, df.tip, margins= True)

In [ ]:
import scipy.stats as stats

data = pd.crosstab(df_tip.p1_products_in_order, df.tip)

stats.chi2_contingency(data)

Da der p-Wert des Tests weniger als 0.05 beträgt wird die Nullhypothese abgelehnt, es gilt H1.

In [ ]:
pd.crosstab(df_tip.p1_products_in_order, df.tip, margins= True, normalize= "all")

In [ ]:
# hji = hi * hj
pd.crosstab(df_tip.p1_products_in_order, df.tip, margins= True, normalize= "all").iloc[1, 1:2].values[0] == 0.447496 * 0.018123

0.447496 * 0.018123 = 0.00811 != 0.009454

**Dadurch ist gezeigt, dass die Tippwahrscheinlichkeit der Bestellungen davon abhängig ist, ob sie die Produkte: Organic Garlic, Banana und Sauvignon Blanc enthält.**  
**Die Bestelungen die die Produkte enthalten, erhalten statistisch gesehen mehr Trinkgeld.** 



### **P2:** Lassen sich regionale Unterschiede im Trinkgeldverhalten erkennen?

In [ ]:
df_p2 = df.copy()

- Durchschnittliche Wahrscheinlichkeit das eine Bestellung Trinkgeld erhält, pro County berechnen

In [ ]:
df_county_tip1 = df_p2.groupby(["county", "order_id"]).agg(tip = ("tip", "mean")).reset_index()
df_county_tip2 = df_county_tip1.groupby("county").agg(tip = ("tip", "mean"), orders= ("order_id", "count")).sort_values(by= "tip", ascending= False).reset_index()
df_county_tip2["orders"] = df_county_tip2.orders / df_county_tip2.orders.sum()

In [ ]:
df_county_tip2.head(10)

**Plotten der Tippwahrscheinlichkeit pro County:**

In [ ]:
# Plot P2: Tippwahrscheinlichkeit pro County

fig1 = make_subplots(cols= 1, rows= 2, shared_xaxes= True, shared_yaxes= False, vertical_spacing= 0)

for county in df_county_tip2.county:

 if df_county_tip2[df_county_tip2.county == county].tip.values >= 0.5:
  color = px.colors.sequential.Purpor[5]
  text= df_county_tip2.loc[df_county_tip2.county == county].tip.values.round(3)

  fig1.add_trace(
   go.Bar(
    x= df_county_tip2.loc[df_county_tip2.county == county].county,
    y= df_county_tip2.loc[df_county_tip2.county == county].tip,
    name= f"{county}: {(df_county_tip2.loc[df_county_tip2.county == county].tip.values[0] * 100).round(2)}%",
    showlegend= True,
    marker= dict(color= color),
    width= 0.8,
    text= text,
    base= 0,
    legendgroup= f"{county}",
    hovertemplate = f"<b>{county}</b><br><br>" +
                "Tipp: %{y}<br>" +
                "<extra></extra>"
   ),
   col= 1,
   row= 1
  )

  fig1.add_trace(
   go.Bar(
    x= df_county_tip2.loc[df_county_tip2.county == county].county,
    y= df_county_tip2.loc[df_county_tip2.county == county].orders,
    name= f"Orders: {(df_county_tip2.loc[df_county_tip2.county == county].orders.values[0] * 100).round(3)}%",
    base = -df_county_tip2.loc[df_county_tip2.county == county].orders,
    showlegend= True,
    marker= dict(color= color),
    width= 0.8,
    legendgroup= f"{county}",
    hovertemplate = f"<b>{county}</b><br><br>" +
                "Orders: %{base}<br>" +
                "<extra></extra>"
   ),
   col= 1,
   row= 2
  )

 elif 0.4 < df_county_tip2.loc[df_county_tip2.county == county].tip.values < 0.5:
  color= px.colors.sequential.Greys[3]

  fig1.add_trace(
   go.Bar(
    x= df_county_tip2.loc[df_county_tip2.county == county].county,
    y= df_county_tip2.loc[df_county_tip2.county == county].tip,
    name= f"{county}: {(df_county_tip2.loc[df_county_tip2.county == county].tip.values[0] * 100).round(2)}%",
    showlegend= True,
    marker= dict(color= color),
    width= 0.8,
    base= 0,
    legendgroup= f"{county}",
    hovertemplate = f"<b>{county}</b><br><br>" +
                "Tipp: %{y}<br>" +
                "<extra></extra>"
   ),
   col= 1,
   row= 1
  )

  fig1.add_trace(
   go.Bar(
    x= df_county_tip2.loc[df_county_tip2.county == county].county,
    y= df_county_tip2.loc[df_county_tip2.county == county].orders,
    name= f"Orders: {(df_county_tip2.loc[df_county_tip2.county == county].orders.values[0] * 100).round(3)}%",
    base = -df_county_tip2.loc[df_county_tip2.county == county].orders,
    showlegend= True,
    marker= dict(color= color),
    width= 0.8,
    legendgroup= f"{county}",
    hovertemplate = f"<b>{county}</b><br><br>" +
                "Orders: %{base}<br>" +
                "<extra></extra>"
   ),
   col= 1,
   row= 2
  )

 elif 0.3 < df_county_tip2.loc[df_county_tip2.county == county].tip.values < 0.4:
  color= px.colors.sequential.Greys[2]

  fig1.add_trace(
   go.Bar(
    x= df_county_tip2.loc[df_county_tip2.county == county].county,
    y= df_county_tip2.loc[df_county_tip2.county == county].tip,
    name= f"{county}: {(df_county_tip2.loc[df_county_tip2.county == county].tip.values[0] * 100).round(2)}%",
    showlegend= True,
    marker= dict(color= color),
    width= 0.8,
    base= 0,
    legendgroup= f"{county}",
    hovertemplate = f"<b>{county}</b><br><br>" +
                "Tipp: %{y}<br>" +
                "<extra></extra>"
   ),
   col= 1,
   row= 1
  )

  fig1.add_trace(
   go.Bar(
    x= df_county_tip2.loc[df_county_tip2.county == county].county,
    y= df_county_tip2.loc[df_county_tip2.county == county].orders,
    name= f"Orders: {(df_county_tip2.loc[df_county_tip2.county == county].orders.values[0] * 100).round(3)}%",
    base = -df_county_tip2.loc[df_county_tip2.county == county].orders,
    showlegend= True,
    marker= dict(color= color),
    width= 0.8,
    legendgroup= f"{county}",
    hovertemplate = f"<b>{county}</b><br><br>" +
                "Orders: %{base}<br>" +
                "<extra></extra>"
   ),
   col= 1,
   row= 2
  )

 elif 0.2 < df_county_tip2.loc[df_county_tip2.county == county].tip.values < 0.3:
  color= px.colors.sequential.Greys[1]
  text= df_county_tip2.loc[df_county_tip2.county == county].tip.values.round(3)

  fig1.add_trace(
   go.Bar(
    x= df_county_tip2.loc[df_county_tip2.county == county].county,
    y= df_county_tip2.loc[df_county_tip2.county == county].tip,
    name= f"{county}: {(df_county_tip2.loc[df_county_tip2.county == county].tip.values[0] * 100).round(2)}%",
    showlegend= True,
    marker= dict(color= color),
    width= 0.8,
    text= text,
    base= 0,
    legendgroup= f"{county}",
    hovertemplate = f"<b>{county}</b><br><br>" +
                "Tipp: %{y}<br>" +
                "<extra></extra>"
   ),
   col= 1,
   row= 1
  )

  fig1.add_trace(
   go.Bar(
    x= df_county_tip2.loc[df_county_tip2.county == county].county,
    y= df_county_tip2.loc[df_county_tip2.county == county].orders,
    name= f"Orders: {(df_county_tip2.loc[df_county_tip2.county == county].orders.values[0] * 100).round(3)}%",
    base = -df_county_tip2.loc[df_county_tip2.county == county].orders,
    showlegend= True,
    marker= dict(color= color),
    width= 0.8,
    legendgroup= f"{county}",
    hovertemplate = f"<b>{county}</b><br><br>" +
                "Orders: %{base}<br>" +
                "<extra></extra>"
   ),
   col= 1,
   row= 2
  )

fig1.add_hline(
 y= 0.5,
 opacity= 0.3,
 line_dash= "solid",
 annotation_text= 0.5,
 annotation_position= "top right",
 line_color= px.colors.sequential.Greys[7],
 annotation_font_color= px.colors.sequential.Greys[7],
 annotation_font_size= 12,
 layer= "below",
 col= 1,
 row= 1
)

fig1.add_hline(
 y= 0.3,
 opacity= 0.3,
 line_dash= "solid",
 annotation_text= 0.3,
 annotation_position= "top right",
 line_color= px.colors.sequential.Greys[7],
 annotation_font_color= px.colors.sequential.Greys[7],
 annotation_font_size= 12,
 layer= "below",
 col= 1,
 row= 1
)

fig1.add_hline(
 y= -0.05,
 opacity= 0.3,
 line_dash= "solid",
 annotation_text= -0.05,
 annotation_position= "top right",
 line_color= px.colors.sequential.Greys[7],
 annotation_font_color= px.colors.sequential.Greys[7],
 annotation_font_size= 12,
 layer= "below",
 col= 1,
 row= 2
)

fig1.add_hline(
 y= -0.03,
 opacity= 0.3,
 line_dash= "solid",
 annotation_text= -0.03,
 annotation_position= "top right",
 line_color= px.colors.sequential.Greys[7],
 annotation_font_color= px.colors.sequential.Greys[7],
 annotation_font_size= 12,
 layer= "below",
 col= 1,
 row= 2
)

fig1.add_hline(
 y= 0,
 opacity= 0.2,
 line_dash= "solid",
 line_color= "black",
 col= 1,
 row= 1
)

fig1.add_hline(
 y= 0,
 opacity= 0.2,
 line_dash= "solid",
 line_color= "black",
 col= 1,
 row= 2
)

fig1.add_vline(
 x= -1,
 opacity= 0.8,
 line_dash= "solid",
 line_color= "black",
 layer= "below",
 col= 1,
 row= 2
)

fig1.add_vline(
 x= -1,
 opacity= 0.8,
 line_dash= "solid",
 line_color= "black",
 layer= "below",
 col= 1,
 row= 1
)

fig1.add_hline(
 y= -0.075,
 opacity= 0.8,
 line_dash= "solid",
 line_color= "black",
 col= 1,
 row= 2
)

fig1.update_annotations(opacity= 0.8)
fig1.update_layout(title= dict(text= "<b>Regionale Unterschiede im Trinkgeldverhalten</b>", font_size= 20, xanchor= "center", yanchor= "top", y=0.935, x=0.475), legend_title_text = "<b>Counties</b>", height= 700)
fig1.update_xaxes(title_text="<b>Counties</b>", ticks= "outside", col= 1, row= 2, showline= True)
fig1.update_xaxes(showline= False, col= 1, row= 1)
fig1.update_yaxes(title_text="<b>Tipp Wshk.</b>", ticks= "outside", showgrid= True, gridcolor= "grey", col= 1, row= 1, tickvals= [0, 0.1, 0.2, 0.3, 0.4, 0.5], showline= False)
fig1.update_yaxes(title_text="<b>Order Anteil</b>", ticks= "outside", showgrid= True, gridcolor= "grey", col= 1, row= 2, tickvals= [0, -0.01, -0.03, -0.05, -0.07])
fig1.update_layout(plot_bgcolor= "white", xaxis= dict(linecolor= "black"), yaxis= dict(linecolor= "black"))
fig1.show()

In [ ]:
with urlopen('https://raw.githubusercontent.com/plotly/datasets/master/geojson-counties-fips.json') as response:
    geojsn = json.load(response)

In [ ]:
a = geojsn["features"].__len__()
for i in range(a,0,-1):
    if geojsn["features"][i-1]["properties"]["STATE"] != "06":
        geojsn["features"].pop(i-1)

In [ ]:
# Plot P2: Regionale Unterschiede im Tippverhalten Map

colorscales = [(0,px.colors.sequential.Greys[0]),
                (0.3, px.colors.sequential.Greys[2]),
                (0.6, px.colors.sequential.Purp[3]),
                (0.8, px.colors.sequential.Purp[5]),
                (1,px.colors.sequential.Purpor[5])]


fig = make_subplots(cols= 1, rows= 3, specs = [[dict(type= "choroplethmapbox", rowspan= 2)],[dict(type= "choroplethmapbox", rowspan= 2)], [dict(type= "table")]], vertical_spacing = 0.03, shared_xaxes= True)

fig.add_trace(
    go.Choroplethmapbox(
        geojson= geojsn,
        featureidkey= "properties.NAME",
        locations= df_county_tip2.county,
        z= df_county_tip2.tip,
        showlegend=False,
        showscale=True,
        marker= dict(line= dict(width= 0.2, color= "black"), opacity= 1),
        colorscale = colorscales
    ),
    col= 1,
    row= 1
)

fig.add_trace(
    go.Table(
        columnorder= [1, 2, 3, 4, 5],
        columnwidth= [1, 1, 1, 1, 1],
        header= dict(
            values= list(df_county_tip2.columns),
            fill_color= px.colors.sequential.Purp[3],
            line_color= px.colors.sequential.Greys[4],
            font= dict(color= "white", size=15),
            height= 35
        ),
        cells= dict(
            values= df_county_tip2.round(3).T.values,
            line_color= px.colors.sequential.Greys[4],
            fill_color= "white",
            font= dict(color= "black", size=13),
            align= "center",
            height= 30
        )
    ),
    col= 1,
    row= 3
)

fig.update_layout(mapbox_style="carto-positron",
                    mapbox_zoom= 4.3,
                    mapbox_center = {"lat": 37, "lon": -120},
                    height= 700,
                    title= dict(text= "<b>Regionale Unterschiede im Trinkgeldverhalten</b>",
                                font_size= 20,
                                y= 0.99,
                                x= 0.45
                            )
                )
fig.update_layout(margin={"r":0,"t":0,"l":0,"b":0})
fig.show()

- Trinkgeldwahrscheinlichkeit bewegt sich im Bereich von 28.2% in Tulare bis 53.1% in San Diego

- Placer und Trinity haben 0.0076% und 0.0305% zu geringen Anteil am Bestelldatensatz, weshalb sie für Aussagen nicht berücksichtig werden

### **P3:** Welche Attribute einer Bestellung wirken sich auf das Trinkgeldverhalten aus?

In [160]:
df_p3 = df.copy()

#### Korrelationsanalyse

In [ ]:
df_p3[["product_id", "add_to_cart_order",	"reordered",	"order_number",	"order_dow",	"order_hour_of_day",	"days_since_prior_order",	"product_name",	"aisle_id",	"department_id",	"county", "tip"]].corr()

- Werte sind alle < 0.15, daher lässt sich über die Korrelation keine ausreichendere Aussage zu der Frage treffen  

#### Machine Learning Modell und Feature Importance:

- Da mit der Korrelation keine richtige Aussage getroffen werden konnte, wird anhand der Feature Importance eines Modelles versucht die Attribute zu finden, die die Trinkgeldwahrscheinlichkeit am stärsten beeinflussen
  
- Dabei werden nur 10% des gesamt Datensatz verwendet, um Rechenzeit einzusparen
  
- Anschließend wird ein Entscheidungsbaum auf diesem Sample trainiert und die Feature Importance der einzelene Attribute des Datensatzes bestimmt und diese dann genauer analysiert

##### Preprocessing

In [ ]:
df_ml = df_p3.reset_index(drop= True)

**Sampling des Datensatzes auf 10% um Rechenzeit zu ersparen**

In [ ]:
df_sample = df_ml.sample(frac= 0.1)

In [ ]:
df_classifier = df_sample.copy()

**One Hot Encoding der Spalten County und days_since_prior_order zum weiterverarbeiten**

In [ ]:
df_classifier = pd.get_dummies(df_classifier, columns= ["county", "days_since_prior_order"], dummy_na= True)

In [ ]:
X = df_classifier.drop(["product_name", "department", "aisle", "tip"], axis= 1)
y = df_classifier["tip"]

In [ ]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size= 0.2, random_state=0)

##### **DTC**

- Mittels einer 5-Fachen Kreuzvalidierung und Hyperparametertuning wird die beste Kombination aus Baumtiefe (max_depth) und minimaler Anzahl an benötigter Beispiele für einen Split (min_samples) gesucht:

In [ ]:
gs_crss_dtc = GridSearchCV(
 estimator= DecisionTreeClassifier(),
 param_grid= dict(min_samples_split= [2, 10, 20, 50, 100], max_depth= [1, 3, 5, 9], random_state= [0]),
 scoring= "accuracy",
 n_jobs= -1
).fit(X_train, y_train)

print(f"Best param: {gs_crss_dtc.best_params_} with acc: {gs_crss_dtc.best_score_}")

In [ ]:
best_dtc = DecisionTreeClassifier(max_depth= 9, min_samples_split= 2, random_state= 0).fit(X_train, y_train)

y_best_train_pred = best_dtc.predict(X_train)
y_best_test_pred = best_dtc.predict(X_test)

print_evaluation("DTC", X_train, X_test, y_train, y_test, y_best_train_pred, y_best_test_pred)

In [ ]:
feature_importances = pd.DataFrame.from_dict(dict(zip(X.columns, best_dtc.feature_importances_)), orient= "index", columns= ["feature_importance"])

df_feature_imp = feature_importances.reset_index().rename(columns= {"index": "feature"})

In [ ]:
df_feature_imp = df_feature_imp.sort_values(by= "feature_importance", ascending= True).tail(9)
df_feature_imp

In [ ]:
# P3 Plot: Feature Importance

fig = go.Figure()


for feature in df_feature_imp.feature.unique():

 if df_feature_imp.loc[df_feature_imp.feature == feature].feature_importance.values >= 0.1:
  color= px.colors.sequential.Purpor[5]

  fig.add_trace(
   go.Bar(
    x= df_feature_imp.loc[df_feature_imp.feature == feature].feature_importance,
    y= df_feature_imp.loc[df_feature_imp.feature == feature].feature,
    marker= dict(color= color),
    orientation = "h",
    text= df_feature_imp.loc[df_feature_imp.feature == feature].feature_importance.round(2),
    showlegend= False,
    width= 0.8,
    hovertemplate= f"<b>{feature}</b><br><br>" +
                   "feature_importance: %{x}<br>" +
                   "<extra></extra>"
    )
  )
 else:
  color= px.colors.sequential.Greys[3]

  fig.add_trace(
   go.Bar(
    x= df_feature_imp.loc[df_feature_imp.feature == feature].feature_importance,
    y= df_feature_imp.loc[df_feature_imp.feature == feature].feature,
    marker= dict(color= color),
    orientation = "h",
    text= df_feature_imp.loc[df_feature_imp.feature == feature].feature_importance.round(3),
    showlegend= False,
    width= 0.8,
    hovertemplate= f"<b>{feature}</b><br><br>" +
                   "feature_importance: %{x}<br>" +
                   "<extra></extra>"
    )
  )

fig.update_layout(title= dict(text= "<b>Feature Importance der Attribute</b>", font_size= 20, xanchor= "center", yanchor= "top", y=0.935, x=0.475), height= 600)
fig.update_xaxes(title_text="<b>Importance</b>", ticks= "outside")
fig.update_xaxes(showline= True)
fig.update_yaxes(title_text="<b>Feature</b>", ticks= "outside", showgrid= True, gridcolor= "grey", showline= True)
fig.update_layout(plot_bgcolor= "white", xaxis= dict(linecolor= "black"), yaxis= dict(linecolor= "black"))
fig.show()

#### Attribut: order_dow

- Gruppieren der Bestelldaten nach Wochentagen und anschließend berechnen der Trinkgeldwahrscheinlichkeit  

In [161]:
df1 = df_p3.groupby(["order_dow", "order_id"], as_index= False).agg(tip_orders_per_day = ("tip", "mean"))
df1.head()

order_dow  order_id  tip_orders_per_day
0          0         4                 1.0
1          0         6                 0.0
2          0         9                 1.0
3          0        10                 0.0
4          0        19                 1.0

In [162]:
df_order_dow= df1.groupby("order_dow", as_index= False).agg(tip = ("tip_orders_per_day","sum"), orders= ("order_id", "nunique"))
df_order_dow["tip_norm"] = df_order_dow.tip / df_order_dow.orders
df_order_dow["orders_norm"] = df_order_dow.orders / df_order_dow.orders.sum()
df_order_dow.head()

order_dow      tip  orders  tip_norm  orders_norm
0          0  53749.0  106842  0.503070     0.176041
1          1  54928.0  104713  0.524558     0.172533
2          2  30614.0   83177  0.368058     0.137049
3          3  27884.0   77329  0.360589     0.127413
4          4  27780.0   75323  0.368812     0.124108

In [163]:
df_order_dow["order_dow_day"] = ["Montag", "Dienstag", "Mittwoch", "Donnerstag", "Freitag", "Samstag", "Sonntag"]
df_order_dow.head()

order_dow      tip  orders  tip_norm  orders_norm order_dow_day
0          0  53749.0  106842  0.503070     0.176041        Montag
1          1  54928.0  104713  0.524558     0.172533      Dienstag
2          2  30614.0   83177  0.368058     0.137049      Mittwoch
3          3  27884.0   77329  0.360589     0.127413    Donnerstag
4          4  27780.0   75323  0.368812     0.124108       Freitag

**Plotten der Tippwahrscheinlichkeit pro Wochentag**

In [168]:
# Plot P3: Order_dow

df_order_dow["tip_norm"] = df_order_dow.tip_norm.round(3)
df_order_dow["orders_norm"] = df_order_dow.orders_norm.round(3)

fig = make_subplots(cols= 1, rows=2, specs= [[dict(type= "table")], [dict(type= "bar")]], vertical_spacing= 0.1)

fig.add_trace(
  go.Table(
    columnorder= [1, 2, 3, 4, 5],
    columnwidth= [1, 1, 1, 1, 1],
    header=dict(
        values=list(df_order_dow.columns),
        fill_color= px.colors.sequential.Purpor[5],
        line_color= px.colors.sequential.Greys[2],
        font= dict(color= "white", size=15),
        height= 35
        ),
    cells=dict(
        values=df_order_dow.T.values,
        fill_color= "white",
        line_color= px.colors.sequential.Greys[2],
        font= dict(color= "black", size=13),
        align= "center",
        height= 30
        )
    ),
    col= 1,
    row= 1
)

for order_dow in df_order_dow.order_dow:

 if df_order_dow[df_order_dow.order_dow == order_dow].tip_norm.values >= 0.5:
  color = px.colors.sequential.Purpor[5]
  text= df_order_dow.loc[df_order_dow.order_dow == order_dow].tip_norm.values.round(3)

  fig.add_trace(
   go.Bar(
    x= df_order_dow.loc[df_order_dow.order_dow == order_dow].order_dow,
    y= df_order_dow.loc[df_order_dow.order_dow == order_dow].tip_norm,
    name= f"Tipp: {(df_order_dow.loc[df_order_dow.order_dow == order_dow].tip_norm.values[0] * 100).round(2)}%",
    showlegend= True,
    marker= dict(color= color),
    width= 0.8,
    text= text,
    base= 0,
    legendgroup= f"{order_dow}",
    legendgrouptitle_text = f"{order_dow}: {df_order_dow.loc[df_order_dow.order_dow == order_dow].order_dow_day.values[0]}",
    hovertemplate = f"<b>{df_order_dow.loc[df_order_dow.order_dow == order_dow].order_dow_day.values[0]}</b><br><br>" +
                "Tipp: %{y}<br>" +
                "<extra></extra>"
   ),
   col= 1,
   row= 2
  )

 else:
  color= px.colors.sequential.Greys[2]

  fig.add_trace(
   go.Bar(
    x= df_order_dow.loc[df_order_dow.order_dow == order_dow].order_dow,
    y= df_order_dow.loc[df_order_dow.order_dow == order_dow].tip_norm,
    name= f"Tipp: {(df_order_dow.loc[df_order_dow.order_dow == order_dow].tip_norm.values[0] * 100).round(2)}%",
    showlegend= True,
    marker= dict(color= color),
    width= 0.8,
    base= 0,
    legendgroup= f"{order_dow}",
    legendgrouptitle_text = f"{order_dow}: {df_order_dow.loc[df_order_dow.order_dow == order_dow].order_dow_day.values[0]}",
    hovertemplate = f"<b>{df_order_dow.loc[df_order_dow.order_dow == order_dow].order_dow_day.values[0]}</b><br><br>" +
                "Tipp: %{y}<br>" +
                "<extra></extra>"
   ),
   col= 1,
   row= 2
  )

fig.update_layout(title= dict(text= "<b>Tippwahrscheinlichkeit nach Wochentagen</b>", font_size= 20, xanchor= "center", yanchor= "top", y=0.935, x=0.475), legend_title_text = "<b>Tage</b>", height= 780)
fig.update_xaxes(title_text="<b>Tage</b>", ticks= "outside", col= 1, row= 2)
fig.update_xaxes(showline= True, col= 1, row= 2)
fig.update_yaxes(title_text="<b>Tipp Wshk.</b>", ticks= "outside", showgrid= True, gridcolor= "grey", col= 1, row= 2, showline= True, tickvals= [0.1, 0.3, 0.5])
fig.update_layout(plot_bgcolor= "white", xaxis= dict(linecolor= "black"), yaxis= dict(linecolor= "black"))
fig.show()

- Die Trinkgeldwahrscheinlichkeit ist am Montagen 50.3% und Dienstagen 52.5% am höchstens

- Im Schnitt ist es damit ca. 14% wahrscheinlicher Trinkgeld an Montagen und Dienstagen zu erhalten, als an den restlichen Wochentagen


#### Attribut: order_number

- Gruppieren der Bestellungen nach order_number pro User und berechnen der Trinkgeldwahrscheinlichkeit für Intervalle der Bestellungsanzahl 

In [169]:
df_order_number1= df_p3.groupby(["user_id", "order_id", "order_number"], as_index= False).agg(tip= ("tip", "mean")).sort_values(by= ["user_id", "order_number"])

In [170]:
df_order_number2 = df_order_number1.groupby("user_id", as_index= False).agg(order_id_max = ("order_number", "max"), tip= ("tip", "sum"), orders= ("order_id", "nunique"))

In [171]:
df_order_number2 = df_order_number1.groupby("user_id", as_index= False).agg(order_id_max = ("order_number", "max"), tip= ("tip", "sum"), orders= ("order_id", "nunique"))

order_bins = [0, 10, 20, 30, 40, 50, 60, 70, 80, 90, 100]
df_order_number2["order_number_cut"] = pd.cut(df_order_number2["order_id_max"], bins= order_bins, include_lowest= False)
df_order_number2.head()

user_id  order_id_max   tip  orders order_number_cut
0        3            12  10.0      12         (10, 20]
1        7            21  12.0      21         (20, 30]
2       14            14   8.0      14         (10, 20]
3       23             5   0.0       5          (0, 10]
4       25             3   2.0       3          (0, 10]

In [172]:
df_order_number3 = df_order_number2.groupby("order_number_cut", as_index= False).agg(users= ("user_id", "nunique"), order_max= ("order_id_max", "nunique"), tip= ("tip", "sum"), orders= ("orders", "sum"))
df_order_number3["tip_norm"] = (df_order_number3.tip / df_order_number3.orders).round(3)
df_order_number3["orders_norm"] = (df_order_number3.orders / df_order_number3.orders.sum()).round(3)
df_order_number3["order_number_cut"] = df_order_number3.order_number_cut.astype("str")
df_order_number3.head()

order_number_cut  users  order_max      tip  orders  tip_norm  orders_norm
0          (0, 10]  19369          8  32250.0  118378     0.272        0.195
1         (10, 20]   8959         10  47207.0  131584     0.359        0.217
2         (20, 30]   3859         10  43187.0   95863     0.451        0.158
3         (30, 40]   2011         10  36084.0   70337     0.513        0.116
4         (40, 50]   1249         10  30464.0   56223     0.542        0.093

**Plotten Trinkgeldwahrscheinlichkeit nach Bestellungsanzahl pro User**

In [178]:
# Plot P3: order_number 

fig = make_subplots(cols= 1, rows=2, specs= [[dict(type= "table")], [dict(type= "bar")]], vertical_spacing= 0.05)

fig.add_trace(
  go.Table(
    columnorder= [1, 2, 3, 4, 5, 6, 7],
    columnwidth= [1, 1, 1, 1, 1, 1, 1],
    header=dict(
        values= list(df_order_number3.columns),
        fill_color= px.colors.sequential.Purpor[5],
        line_color= px.colors.sequential.Greys[2],
        font= dict(color= "white", size=15),
        height= 35
        ),
    cells=dict(
        values= df_order_number3.T.values,
        fill_color= "white",
        line_color= px.colors.sequential.Greys[2],
        font= dict(color= "black", size=13),
        align= "center",
        height= 30
        )
    ),
    col= 1,
    row= 1
)


for interval in df_order_number3.order_number_cut.unique():

 if 0.5 <= df_order_number3.loc[df_order_number3.order_number_cut == interval].tip_norm.values:
  color = px.colors.sequential.Purpor[5]

  fig.add_trace(
   go.Bar(
    x= df_order_number3.loc[df_order_number3.order_number_cut == interval].order_number_cut.values,
    y= df_order_number3.loc[df_order_number3.order_number_cut == interval].tip_norm.values,
    name= f"Tipp%: {(df_order_number3.loc[df_order_number3.order_number_cut == interval].tip_norm.values[0] * 100).round(2)}%",
    showlegend= True,
    marker= dict(color= color),
    width= 0.8,
    legendgroup= f"{interval}",
    legendgrouptitle_text = f"{interval}",
    hovertemplate = f"<b>{interval}</b><br><br>" +
        "Tipp: %{y:.0%}<br>" +
        "<extra></extra>",
   ),
  )

 else:
  color = px.colors.sequential.Greys[3]

  fig.add_trace(
   go.Bar(
    x= df_order_number3.loc[df_order_number3.order_number_cut == interval].order_number_cut.values,
    y= df_order_number3.loc[df_order_number3.order_number_cut == interval].tip_norm.values,
    name= f"Tipp%: {(df_order_number3.loc[df_order_number3.order_number_cut == interval].tip_norm.values[0] * 100).round(2)}%",
    showlegend= True,
    marker= dict(color= color),
    width= 0.8,
    legendgroup= f"{interval}",
    legendgrouptitle_text = f"{interval}",
    hovertemplate = f"<b>{interval}</b><br><br>" +
        "Tipp: %{y:.0%}<br>" +
        "<extra></extra>",
   ),
  )

fig.update_layout(title= dict(text= "<b>Tippwahrscheinlichkeit nach Bestellungsanzahl pro User</b>", font_size= 20, xanchor= "center", yanchor= "top", y=0.935, x=0.475), legend_title_text = "<b>Bestellungen</b>", height= 800)
fig.update_xaxes(title_text="<b>Bestellungsanzahl in Intervallen</b>", ticks= "outside", col= 1, row= 2, showline= True)
fig.update_yaxes(title_text="<b>Tipp Wshk.</b>", ticks= "outside", showgrid= True, gridcolor= "grey", col= 1, row= 2, showline= True, tickvals= [0, 0.25, 0.5, 0.75])
fig.update_layout(plot_bgcolor= "white", xaxis= dict(linecolor= "black"), yaxis= dict(linecolor= "black"))

fig.show()

- Bestellungen von Usern mit einer Bestellungsanzahl von 40 - 60 Bestellungen haben mit rund 54% die höchste Trinkgeldwahrscheinlichkeit

#### Attribut: order_hour_of_day

- Gruppieren der Bestelldaten nach Tagesstunden und berechnen der Trinkgeldwahrscheinlichkeit 

In [ ]:
df_h = df_p3.groupby(["order_hour_of_day", "order_id"], as_index= False).agg(tip= ("tip", "mean"))
df_hour = df_h.groupby(["order_hour_of_day"], as_index= False).agg(orders= ("order_id", "nunique"), tip= ("tip", "sum"))
df_hour["tip_norm"] = df_hour.tip / df_hour.orders
df_hour["orders_norm"] = df_hour.orders / df_hour.orders.sum()
df_hour.head()

**Plotten der Tippwahrscheinlichkeit pro Stunde**

In [ ]:
# Plot P3: Order_hour_of_day

fig1 = make_subplots(cols= 1, rows= 2, shared_xaxes= True, shared_yaxes= False, vertical_spacing= 0)

for order_hour_of_day in df_hour.order_hour_of_day:

 if df_hour[df_hour.order_hour_of_day == order_hour_of_day].tip_norm.values >= 0.5:
  color = px.colors.sequential.Purpor[5]
  text= df_hour.loc[df_hour.order_hour_of_day == order_hour_of_day].tip_norm.values.round(3)

  fig1.add_trace(
   go.Bar(
    x= df_hour.loc[df_hour.order_hour_of_day == order_hour_of_day].order_hour_of_day,
    y= df_hour.loc[df_hour.order_hour_of_day == order_hour_of_day].tip_norm,
    name= f"{order_hour_of_day}: {(df_hour.loc[df_hour.order_hour_of_day == order_hour_of_day].tip_norm.values[0] * 100).round(2)}%",
    showlegend= True,
    marker= dict(color= color),
    width= 0.8,
    text= text,
    base= 0,
    legendgroup= f"{order_hour_of_day}",
    hovertemplate = f"<b>{df_hour.loc[df_hour.order_hour_of_day == order_hour_of_day].order_hour_of_day.values[0]}</b><br><br>" +
                "Tipp: %{y}<br>" +
                "<extra></extra>"
   ),
   col= 1,
   row= 1
  )

  fig1.add_trace(
   go.Bar(
    x= df_hour.loc[df_hour.order_hour_of_day == order_hour_of_day].order_hour_of_day,
    y= df_hour.loc[df_hour.order_hour_of_day == order_hour_of_day].orders_norm,
    name= f"Orders: {(df_hour.loc[df_hour.order_hour_of_day == order_hour_of_day].orders_norm.values[0] * 100).round(3)}%",
    base = -df_hour.loc[df_hour.order_hour_of_day == order_hour_of_day].orders_norm,
    showlegend= True,
    marker= dict(color= color),
    width= 0.8,
    legendgroup= f"{order_hour_of_day}",
    hovertemplate = f"<b>{df_hour.loc[df_hour.order_hour_of_day == order_hour_of_day].order_hour_of_day.values[0]}</b><br><br>" +
                "Tipp: %{base}<br>" +
                "<extra></extra>"
   ),
   col= 1,
   row= 2
  )

 elif 0.4 < df_hour.loc[df_hour.order_hour_of_day == order_hour_of_day].tip_norm.values < 0.5:
  color= px.colors.sequential.Greys[3]

  fig1.add_trace(
   go.Bar(
    x= df_hour.loc[df_hour.order_hour_of_day == order_hour_of_day].order_hour_of_day,
    y= df_hour.loc[df_hour.order_hour_of_day == order_hour_of_day].tip_norm,
    name= f"{order_hour_of_day}: {(df_hour.loc[df_hour.order_hour_of_day == order_hour_of_day].tip_norm.values[0] * 100).round(2)}%",
    showlegend= True,
    marker= dict(color= color),
    width= 0.8,
    base= 0,
    legendgroup= f"{order_hour_of_day}",
    hovertemplate = f"<b>{df_hour.loc[df_hour.order_hour_of_day == order_hour_of_day].order_hour_of_day.values[0]}</b><br><br>" +
                "Tipp: %{y}<br>" +
                "<extra></extra>"
   ),
   col= 1,
   row= 1
  )

  fig1.add_trace(
   go.Bar(
    x= df_hour.loc[df_hour.order_hour_of_day == order_hour_of_day].order_hour_of_day,
    y= df_hour.loc[df_hour.order_hour_of_day == order_hour_of_day].orders_norm,
    name= f"Orders: {(df_hour.loc[df_hour.order_hour_of_day == order_hour_of_day].orders_norm.values[0] * 100).round(3)}%",
    base = -df_hour.loc[df_hour.order_hour_of_day == order_hour_of_day].orders_norm,
    showlegend= True,
    marker= dict(color= color),
    width= 0.8,
    legendgroup= f"{order_hour_of_day}",
    hovertemplate = f"<b>{df_hour.loc[df_hour.order_hour_of_day == order_hour_of_day].order_hour_of_day.values[0]}</b><br><br>" +
                "Tipp: %{base}<br>" +
                "<extra></extra>"
   ),
   col= 1,
   row= 2
  )

 elif 0.3 < df_hour.loc[df_hour.order_hour_of_day == order_hour_of_day].tip_norm.values < 0.4:
  color= px.colors.sequential.Greys[2]

  fig1.add_trace(
   go.Bar(
    x= df_hour.loc[df_hour.order_hour_of_day == order_hour_of_day].order_hour_of_day,
    y= df_hour.loc[df_hour.order_hour_of_day == order_hour_of_day].tip_norm,
    name= f"{order_hour_of_day}: {(df_hour.loc[df_hour.order_hour_of_day == order_hour_of_day].tip_norm.values[0] * 100).round(2)}%",
    showlegend= True,
    marker= dict(color= color),
    width= 0.8,
    base= 0,
    legendgroup= f"{order_hour_of_day}",
    hovertemplate = f"<b>{df_hour.loc[df_hour.order_hour_of_day == order_hour_of_day].order_hour_of_day.values[0]}</b><br><br>" +
                "Tipp: %{y}<br>" +
                "<extra></extra>"
   ),
   col= 1,
   row= 1
  )

  fig1.add_trace(
   go.Bar(
    x= df_hour.loc[df_hour.order_hour_of_day == order_hour_of_day].order_hour_of_day,
    y= df_hour.loc[df_hour.order_hour_of_day == order_hour_of_day].orders_norm,
    name= f"Orders: {(df_hour.loc[df_hour.order_hour_of_day == order_hour_of_day].orders_norm.values[0] * 100).round(3)}%",
    base = -df_hour.loc[df_hour.order_hour_of_day == order_hour_of_day].orders_norm,
    showlegend= True,
    marker= dict(color= color),
    width= 0.8,
    legendgroup= f"{order_hour_of_day}",
    hovertemplate = f"<b>{df_hour.loc[df_hour.order_hour_of_day == order_hour_of_day].order_hour_of_day.values[0]}</b><br><br>" +
                "Tipp: %{base}<br>" +
                "<extra></extra>"
   ),
   col= 1,
   row= 2
  )

fig1.add_hline(
 y= 0,
 opacity= 0.3,
 line_dash= "solid",
 line_color= "grey",
 col= 1,
 row= 1
)

fig1.add_hline(
 y= 0,
 opacity= 0.3,
 line_dash= "solid",
 line_color= "grey",
 col= 1,
 row= 2
)

fig1.add_vline(
 x= -1,
 opacity= 0.8,
 line_dash= "solid",
 line_color= "black",
 col= 1,
 row= 2
)

fig1.add_vline(
 x= -1,
 opacity= 0.8,
 line_dash= "solid",
 line_color= "black",
 col= 1,
 row= 1
)

fig1.add_hline(
 y= -0.1,
 opacity= 0.8,
 line_dash= "solid",
 line_color= "black",
 col= 1,
 row= 2
)

fig1.update_annotations(opacity= 0.8)
fig1.update_layout(title= dict(text= "<b>Tippwahrscheinlichkeit nach Tagesstunden</b>", font_size= 20, xanchor= "center", yanchor= "top", y=0.935, x=0.475), legend_title_text = "<b>Stunden</b>", height= 700)
fig1.update_xaxes(title_text="<b>Stunden</b>", ticks= "outside", col= 1, row= 2)
fig1.update_xaxes(showline= False, col= 1, row= 1)
fig1.update_yaxes(title_text="<b>Tipp Wshk.</b>", ticks= "outside", showgrid= True, gridcolor= "grey", col= 1, row= 1, showline= False)
fig1.update_yaxes(title_text="<b>Order Anteil</b>", ticks= "outside", showgrid= True, gridcolor= "grey", col= 1, row= 2, tickvals= [0, -0.025, -0.05, -0.075, -0.1])
fig1.update_layout(plot_bgcolor= "white", xaxis= dict(linecolor= "black"), yaxis= dict(linecolor= "black"))
fig1.show()

- Im Zeitraum von 5 bis 18 Uhr liegt die Trinkgeldwahrscheinlichtkeit ausnahmslos unterhalb von 50%
 
- Außerhalb der "normalen" Arbeitszeit, von 19 bis 4 Uhr, liegt die Trinkgeldwahrscheinlichkeit außer aum 4 Uhr immer über 50%
 
- Bestellungen die um 21 zugestellt wurden haben mit 54.6% die höchste Wahrscheinlichkeit Trinkgeld zu erhalten 

#### Attribut: days_since_prior_order

- Gruppieren der Bestelldaten nach days_since_prior_order und berechnen der Tippwahrscheinlichkeit

In [ ]:
df_days = df_p3.groupby(["days_since_prior_order", "order_id"], as_index=False).agg(tip_mean= ("tip", "mean"))
df_days_since_prior_order= df_days.groupby("days_since_prior_order", as_index= False).agg(tip= ("tip_mean", "sum"), orders= ("order_id", "nunique"))
df_days_since_prior_order["tip_norm"] = df_days_since_prior_order.tip / df_days_since_prior_order.orders
df_days_since_prior_order["orders_norm"] = df_days_since_prior_order.orders / df_days_since_prior_order.orders.sum()
df_days_since_prior_order.head()

**Plotten der Trinkgeldwahrscheinlichkeit anhand wie viele Tage die vorherige Bestellungen zurückliegt**

In [ ]:
# Plot P3: Days_since_prior_order
fig1 = make_subplots(cols= 1, rows= 2, shared_xaxes= True, shared_yaxes= False, vertical_spacing= 0)

for days_since_prior_order in df_days_since_prior_order.days_since_prior_order:

 if df_days_since_prior_order[df_days_since_prior_order.days_since_prior_order == days_since_prior_order].tip_norm.values >= 0.5:
  color = px.colors.sequential.Purpor[5]
  text= df_days_since_prior_order.loc[df_days_since_prior_order.days_since_prior_order == days_since_prior_order].tip_norm.values.round(2)

  fig1.add_trace(
   go.Bar(
    x= df_days_since_prior_order.loc[df_days_since_prior_order.days_since_prior_order == days_since_prior_order].days_since_prior_order,
    y= df_days_since_prior_order.loc[df_days_since_prior_order.days_since_prior_order == days_since_prior_order].tip_norm,
    name= f"{int(days_since_prior_order)}: {(df_days_since_prior_order.loc[df_days_since_prior_order.days_since_prior_order == days_since_prior_order].tip_norm.values[0] * 100).round(2)}%",
    showlegend= True,
    marker= dict(color= color),
    width= 0.8,
    text= text,
    base= 0,
    legendgroup= f"{days_since_prior_order}",
    hovertemplate = f"<b>{df_days_since_prior_order.loc[df_days_since_prior_order.days_since_prior_order == days_since_prior_order].days_since_prior_order.values[0]}</b><br><br>" +
                "Tipp: %{y}<br>" +
                "<extra></extra>"
   ),
   col= 1,
   row= 1
  )

  fig1.add_trace(
   go.Bar(
    x= df_days_since_prior_order.loc[df_days_since_prior_order.days_since_prior_order == days_since_prior_order].days_since_prior_order,
    y= df_days_since_prior_order.loc[df_days_since_prior_order.days_since_prior_order == days_since_prior_order].orders_norm,
    name= f"Orders: {(df_days_since_prior_order.loc[df_days_since_prior_order.days_since_prior_order == days_since_prior_order].orders_norm.values[0] * 100).round(3)}%",
    base = -df_days_since_prior_order.loc[df_days_since_prior_order.days_since_prior_order == days_since_prior_order].orders_norm,
    showlegend= True,
    marker= dict(color= color),
    width= 0.8,
    legendgroup= f"{days_since_prior_order}",
    hovertemplate = f"<b>{df_days_since_prior_order.loc[df_days_since_prior_order.days_since_prior_order == days_since_prior_order].days_since_prior_order.values[0]}</b><br><br>" +
                "Tipp: %{y}<br>" +
                "<extra></extra>"
   ),
   col= 1,
   row= 2
  )

 elif 0.4 < df_days_since_prior_order.loc[df_days_since_prior_order.days_since_prior_order == days_since_prior_order].tip_norm.values < 0.5:
  color= px.colors.sequential.Greys[3]

  fig1.add_trace(
   go.Bar(
    x= df_days_since_prior_order.loc[df_days_since_prior_order.days_since_prior_order == days_since_prior_order].days_since_prior_order,
    y= df_days_since_prior_order.loc[df_days_since_prior_order.days_since_prior_order == days_since_prior_order].tip_norm,
    name= f"{int(days_since_prior_order)}: {(df_days_since_prior_order.loc[df_days_since_prior_order.days_since_prior_order == days_since_prior_order].tip_norm.values[0] * 100).round(2)}%",
    showlegend= True,
    marker= dict(color= color),
    width= 0.8,
    base= 0,
    legendgroup= f"{days_since_prior_order}",
    hovertemplate = f"<b>{df_days_since_prior_order.loc[df_days_since_prior_order.days_since_prior_order == days_since_prior_order].days_since_prior_order.values[0]}</b><br><br>" +
                "Tipp: %{y}<br>" +
                "<extra></extra>"
   ),
   col= 1,
   row= 1
  )

  fig1.add_trace(
   go.Bar(
    x= df_days_since_prior_order.loc[df_days_since_prior_order.days_since_prior_order == days_since_prior_order].days_since_prior_order,
    y= df_days_since_prior_order.loc[df_days_since_prior_order.days_since_prior_order == days_since_prior_order].orders_norm,
    name= f"Orders: {(df_days_since_prior_order.loc[df_days_since_prior_order.days_since_prior_order == days_since_prior_order].orders_norm.values[0] * 100).round(3)}%",
    base = -df_days_since_prior_order.loc[df_days_since_prior_order.days_since_prior_order == days_since_prior_order].orders_norm,
    showlegend= True,
    marker= dict(color= color),
    width= 0.8,
    legendgroup= f"{days_since_prior_order}",
    hovertemplate = f"<b>{df_days_since_prior_order.loc[df_days_since_prior_order.days_since_prior_order == days_since_prior_order].days_since_prior_order.values[0]}</b><br><br>" +
                "Tipp: %{y}<br>" +
                "<extra></extra>"
   ),
   col= 1,
   row= 2
  )

 elif 0.3 < df_days_since_prior_order.loc[df_days_since_prior_order.days_since_prior_order == days_since_prior_order].tip_norm.values < 0.4:
  color= px.colors.sequential.Greys[2]

  fig1.add_trace(
   go.Bar(
    x= df_days_since_prior_order.loc[df_days_since_prior_order.days_since_prior_order == days_since_prior_order].days_since_prior_order,
    y= df_days_since_prior_order.loc[df_days_since_prior_order.days_since_prior_order == days_since_prior_order].tip_norm,
    name= f"{int(days_since_prior_order)}: {(df_days_since_prior_order.loc[df_days_since_prior_order.days_since_prior_order == days_since_prior_order].tip_norm.values[0] * 100).round(2)}%",
    showlegend= True,
    marker= dict(color= color),
    width= 0.8,
    base= 0,
    legendgroup= f"{days_since_prior_order}",
    hovertemplate = f"<b>{df_days_since_prior_order.loc[df_days_since_prior_order.days_since_prior_order == days_since_prior_order].days_since_prior_order.values[0]}</b><br><br>" +
                "Tipp: %{y}<br>" +
                "<extra></extra>"
   ),
   col= 1,
   row= 1
  )

  fig1.add_trace(
   go.Bar(
    x= df_days_since_prior_order.loc[df_days_since_prior_order.days_since_prior_order == days_since_prior_order].days_since_prior_order,
    y= df_days_since_prior_order.loc[df_days_since_prior_order.days_since_prior_order == days_since_prior_order].orders_norm,
    name= f"Orders: {(df_days_since_prior_order.loc[df_days_since_prior_order.days_since_prior_order == days_since_prior_order].orders_norm.values[0] * 100).round(3)}%",
    base = -df_days_since_prior_order.loc[df_days_since_prior_order.days_since_prior_order == days_since_prior_order].orders_norm,
    showlegend= True,
    marker= dict(color= color),
    width= 0.8,
    legendgroup= f"{days_since_prior_order}",
    hovertemplate = f"<b>{df_days_since_prior_order.loc[df_days_since_prior_order.days_since_prior_order == days_since_prior_order].days_since_prior_order.values[0]}</b><br><br>" +
                "Tipp: %{y}<br>" +
                "<extra></extra>"
   ),
   col= 1,
   row= 2
  )

 elif 0.2 < df_days_since_prior_order.loc[df_days_since_prior_order.days_since_prior_order == days_since_prior_order].tip_norm.values < 0.3:
  color= px.colors.sequential.Greys[1]
  text= df_days_since_prior_order.loc[df_days_since_prior_order.days_since_prior_order == days_since_prior_order].tip_norm.values.round(2)

  fig1.add_trace(
   go.Bar(
    x= df_days_since_prior_order.loc[df_days_since_prior_order.days_since_prior_order == days_since_prior_order].days_since_prior_order,
    y= df_days_since_prior_order.loc[df_days_since_prior_order.days_since_prior_order == days_since_prior_order].tip_norm,
    name= f"{int(days_since_prior_order)}: {(df_days_since_prior_order.loc[df_days_since_prior_order.days_since_prior_order == days_since_prior_order].tip_norm.values[0] * 100).round(2)}%",
    showlegend= True,
    marker= dict(color= color),
    width= 0.8,
    text= text,
    base= 0,
    legendgroup= f"{days_since_prior_order}",
    hovertemplate = f"<b>{df_days_since_prior_order.loc[df_days_since_prior_order.days_since_prior_order == days_since_prior_order].days_since_prior_order.values[0]}</b><br><br>" +
                "Tipp: %{y}<br>" +
                "<extra></extra>"
   ),
   col= 1,
   row= 1
  )

  fig1.add_trace(
   go.Bar(
    x= df_days_since_prior_order.loc[df_days_since_prior_order.days_since_prior_order == days_since_prior_order].days_since_prior_order,
    y= df_days_since_prior_order.loc[df_days_since_prior_order.days_since_prior_order == days_since_prior_order].orders_norm,
    name= f"Orders: {(df_days_since_prior_order.loc[df_days_since_prior_order.days_since_prior_order == days_since_prior_order].orders_norm.values[0] * 100).round(3)}%",
    base = -df_days_since_prior_order.loc[df_days_since_prior_order.days_since_prior_order == days_since_prior_order].orders_norm,
    showlegend= True,
    marker= dict(color= color),
    width= 0.8,
    legendgroup= f"{days_since_prior_order}",
    hovertemplate = f"<b>{df_days_since_prior_order.loc[df_days_since_prior_order.days_since_prior_order == days_since_prior_order].days_since_prior_order.values[0]}</b><br><br>" +
                "Tipp: %{y}<br>" +
                "<extra></extra>"
   ),
   col= 1,
   row= 2
  )


fig1.add_hline(
 y= 0,
 opacity= 0.3,
 line_dash= "solid",
 line_color= "grey",
 col= 1,
 row= 1
)

fig1.add_hline(
 y= 0,
 opacity= 0.3,
 line_dash= "solid",
 line_color= "grey",
 col= 1,
 row= 2
)

fig1.add_vline(
 x= -1,
 opacity= 0.8,
 line_dash= "solid",
 line_color= "black",
 col= 1,
 row= 2
)

fig1.add_vline(
 x= -1,
 opacity= 0.8,
 line_dash= "solid",
 line_color= "black",
 col= 1,
 row= 1
)

fig1.add_hline(
 y= -0.18,
 opacity= 0.8,
 line_dash= "solid",
 line_color= "black",
 col= 1,
 row= 2
)

fig1.update_annotations(opacity= 0.8)
fig1.update_layout(title= dict(text= "<b>Tippwahrscheinlichkeit von Bestellungen mit Tagen seit der letzten Bestellungen</b>", font_size= 20, xanchor= "center", yanchor= "top", y=0.935, x=0.475), legend_title_text = "<b>Tage</b>", height= 700)
fig1.update_xaxes(title_text="<b>Tage</b>", ticks= "outside", col= 1, row= 2)
fig1.update_xaxes(showline= False, col= 1, row= 1)
fig1.update_yaxes(title_text="<b>Tipp Wshk.</b>", ticks= "outside", showgrid= True, gridcolor= "grey", col= 1, row= 1, showline= False)
fig1.update_yaxes(title_text="<b>Order Anteil</b>", ticks= "outside", showgrid= True, gridcolor= "grey", col= 1, row= 2)
fig1.update_layout(plot_bgcolor= "white", xaxis= dict(linecolor= "black"), yaxis= dict(linecolor= "black"))
fig1.show()

- Bestellungen die 7 Tage nach der vorherigen erfolgen haben die höchste Trinkgeldwahrscheinlichkeit

- Die Tippwahrscheinlichtkeit sinkt je mit der Anzahl an Tagen die zwischen den Bestellungen liegen, steigt jedoch periodisch alle 7 Tage wieder leicht an 

#### Attribut: department

- Gruppieren der Produkte jeder Bestellung nachedem aus welchem department sie stammen   

In [ ]:
df_dep = df_p3.groupby(["department", "order_id"], as_index= False).agg(tip_mean= ("tip", "mean"), all= ("order_id", "count"))
df_dep1 = df_dep.groupby("department", as_index= False).agg(tip = ("tip_mean", "sum"), orders= ("order_id", "nunique"))
df_dep1["tip_norm"] = df_dep1.tip / df_dep1.orders
df_dep1["orders_norm"] = df_dep1.orders / df_dep1.orders.sum()
df_dep2 = df_dep1.sort_values(by= "tip_norm", ascending= False)

In [ ]:
df_dep2.head()

**Plotten der Trinkgeldwahrscheinlichkeit der Departments**

In [ ]:
# Plot P3: Department

fig1 = make_subplots(cols= 1, rows= 2, shared_xaxes= True, shared_yaxes= False, vertical_spacing= 0)

for department in df_dep2.department:

 if df_dep2[df_dep2.department == department].tip_norm.values >= 0.5:
  color = px.colors.sequential.Purpor[5]
  text= df_dep2.loc[df_dep2.department == department].tip_norm.values.round(3)

  fig1.add_trace(
   go.Bar(
    x= df_dep2.loc[df_dep2.department == department].department,
    y= df_dep2.loc[df_dep2.department == department].tip_norm,
    name= f"Tipp: {(df_dep2.loc[df_dep2.department == department].tip_norm.values[0] * 100).round(2)}%",
    showlegend= True,
    marker= dict(color= color),
    width= 0.8,
    text= text,
    base= 0,
    legendgroup= f"{department}",
    legendgrouptitle_text = f"{department}",
    hovertemplate = f"<b>{df_dep2.loc[df_dep2.department == department].department.values[0]}</b><br><br>" +
                "Tipp: %{y}<br>" +
                "<extra></extra>"
   ),
   col= 1,
   row= 1
  )

  fig1.add_trace(
   go.Bar(
    x= df_dep2.loc[df_dep2.department == department].department,
    y= df_dep2.loc[df_dep2.department == department].orders_norm,
    name= f"Orders: {(df_dep2.loc[df_dep2.department == department].orders_norm.values[0] * 100).round(3)}%",
    base = -df_dep2.loc[df_dep2.department == department].orders_norm,
    showlegend= True,
    marker= dict(color= color),
    width= 0.8,
    legendgroup= f"{department}",
    legendgrouptitle_text = f"{department}",
    hovertemplate = f"<b>{df_dep2.loc[df_dep2.department == department].department.values[0]}</b><br><br>" +
                "Orders: %{base}<br>" +
                "<extra></extra>"
   ),
   col= 1,
   row= 2
  )

 elif 0.4 < df_dep2.loc[df_dep2.department == department].tip_norm.values < 0.5:
  color= px.colors.sequential.Greys[3]

  fig1.add_trace(
   go.Bar(
    x= df_dep2.loc[df_dep2.department == department].department,
    y= df_dep2.loc[df_dep2.department == department].tip_norm,
    name= f"Tipp: {(df_dep2.loc[df_dep2.department == department].tip_norm.values[0] * 100).round(2)}%",
    showlegend= True,
    marker= dict(color= color),
    width= 0.8,
    base= 0,
    legendgroup= f"{department}",
    legendgrouptitle_text = f"{department}",
    hovertemplate = f"<b>{df_dep2.loc[df_dep2.department == department].department.values[0]}</b><br><br>" +
                "Tipp: %{y}<br>" +
                "<extra></extra>"
   ),
   col= 1,
   row= 1
  )

  fig1.add_trace(
   go.Bar(
    x= df_dep2.loc[df_dep2.department == department].department,
    y= df_dep2.loc[df_dep2.department == department].orders_norm,
    name= f"Orders: {(df_dep2.loc[df_dep2.department == department].orders_norm.values[0] * 100).round(3)}%",
    base = -df_dep2.loc[df_dep2.department == department].orders_norm,
    showlegend= True,
    marker= dict(color= color),
    width= 0.8,
    legendgroup= f"{department}",
    legendgrouptitle_text = f"{department}",
    hovertemplate = f"<b>{df_dep2.loc[df_dep2.department == department].department.values[0]}</b><br><br>" +
                "Orders: %{base}<br>" +
                "<extra></extra>"
   ),
   col= 1,
   row= 2
  )

 elif 0.3 < df_dep2.loc[df_dep2.department == department].tip_norm.values < 0.4:
  color= px.colors.sequential.Greys[2]

  fig1.add_trace(
   go.Bar(
    x= df_dep2.loc[df_dep2.department == department].department,
    y= df_dep2.loc[df_dep2.department == department].tip_norm,
    name= f"Tipp: {(df_dep2.loc[df_dep2.department == department].tip_norm.values[0] * 100).round(2)}%",
    showlegend= True,
    marker= dict(color= color),
    width= 0.8,
    base= 0,
    legendgroup= f"{department}",
    legendgrouptitle_text = f"{department}",
    hovertemplate = f"<b>{df_dep2.loc[df_dep2.department == department].department.values[0]}</b><br><br>" +
                "Tipp: %{y}<br>" +
                "<extra></extra>"
   ),
   col= 1,
   row= 1
  )

  fig1.add_trace(
   go.Bar(
    x= df_dep2.loc[df_dep2.department == department].department,
    y= df_dep2.loc[df_dep2.department == department].orders_norm,
    name= f"Orders: {(df_dep2.loc[df_dep2.department == department].orders_norm.values[0] * 100).round(3)}%",
    base = -df_dep2.loc[df_dep2.department == department].orders_norm,
    showlegend= True,
    marker= dict(color= color),
    width= 0.8,
    legendgroup= f"{department}",
    legendgrouptitle_text = f"{department}",
    hovertemplate = f"<b>{df_dep2.loc[df_dep2.department == department].department.values[0]}</b><br><br>" +
                "Orders: %{base}<br>" +
                "<extra></extra>"
   ),
   col= 1,
   row= 2
  )

fig1.add_hline(
 y= 0,
 opacity= 0.3,
 line_dash= "solid",
 line_color= "grey",
 col= 1,
 row= 1
)

fig1.add_hline(
 y= 0,
 opacity= 0.3,
 line_dash= "solid",
 line_color= "grey",
 col= 1,
 row= 2
)

fig1.add_vline(
 x= -1,
 opacity= 0.8,
 line_dash= "solid",
 line_color= "black",
 col= 1,
 row= 2
)

fig1.add_vline(
 x= -1,
 opacity= 0.8,
 line_dash= "solid",
 line_color= "black",
 col= 1,
 row= 1
)

fig1.add_hline(
 y= -0.18,
 opacity= 0.8,
 line_dash= "solid",
 line_color= "black",
 col= 1,
 row= 2
)

fig1.update_annotations(opacity= 0.8)
fig1.update_layout(title= dict(text= "<b>Tippwahrscheinlichkeit der Departments</b>", font_size= 20, xanchor= "center", yanchor= "top", y=0.935, x=0.475), legend_title_text = "<b>Departments</b>", height= 700)
fig1.update_xaxes(title_text="<b>Departments</b>", ticks= "outside", col= 1, row= 2)
fig1.update_xaxes(showline= False, col= 1, row= 1)
fig1.update_yaxes(title_text="<b>Tipp Wshk.</b>", ticks= "outside", showgrid= True, gridcolor= "grey", col= 1, row= 1, showline= False)
fig1.update_yaxes(title_text="<b>Order Anteil</b>", ticks= "outside", showgrid= True, gridcolor= "grey", col= 1, row= 2)
fig1.update_layout(plot_bgcolor= "white", xaxis= dict(linecolor= "black"), yaxis= dict(linecolor= "black"))
fig1.show()

- Bestellungen die Produkte aus den Departments alcohol, breakfast, babies, bulk und snacks enthalten haben eine Trinkgeldwahrscheinlichkeit > 50%

#### Attribut: add_to_cart_order

- Gruppieren der Bestellungen nach ihrer Bestellgröße und anschließend unterteilen in Intervalle

In [ ]:
df_1 = df.groupby(["order_id"], as_index= False).agg(order_size= ("add_to_cart_order", "max"), tip= ("tip", "mean"))
df_2 = df_1.groupby(["order_size"], as_index= False).agg(tip= ("tip", "sum"), orders= ("order_id", "nunique"))
df_2["tip_norm"] = df_2.tip / df_2.orders
df_2["orders_norm"] = df_2.orders / df_2.orders.sum()
df_2.head()

In [ ]:
ordersize_bins = [1, 5, 10, 15, 20, 25, 30, 35, 150]
df_2["order_size_cut"] = pd.cut(df_2["order_size"], bins= ordersize_bins, include_lowest= True)

In [ ]:
df_2_cut = df_2.groupby(["order_size_cut"], as_index= False).agg(tip= ("tip", "sum"), tip_norm= ("tip_norm", "mean"), orders= ("orders", "sum"),)
df_2_cut["orders_norm"] = (df_2_cut.orders / df_2_cut.orders.sum()).round(5)
df_2_cut["tip_norm"] = df_2_cut.tip_norm.round(3)
df_2_cut["order_size_cut"] = df_2_cut.order_size_cut.astype("str")
df_2_cut

**Plotten der Trinkgeldwahrscheinlichkeit anhand der Bestellgröße**

In [ ]:
# Plot P3: Order_size

fig = make_subplots(cols= 1, rows=2, specs= [[dict(type= "table")], [dict(type= "bar")]], vertical_spacing= 0.03)

fig.add_trace(
  go.Table(
    columnorder= [1, 2, 3, 4, 5],
    columnwidth= [1, 1, 1, 1, 1],
    header=dict(
        values= list(df_2_cut.columns),
        fill_color= px.colors.sequential.Purpor[5],
        line_color= px.colors.sequential.Greys[2],
        font= dict(color= "white", size=15),
        height= 35
        ),
    cells=dict(
        values= df_2_cut.T.values,
        fill_color= "white",
        line_color= px.colors.sequential.Greys[2],
        font= dict(color= "black", size=13),
        align= "center",
        height= 30
        )
    ),
    col= 1,
    row= 1
)


for interval in df_2_cut.order_size_cut.unique():

 if df_2_cut.loc[df_2_cut.order_size_cut == interval].tip_norm.values >= 0.5:
  color = px.colors.sequential.Purpor[5]
  text= df_2_cut.loc[df_2_cut.order_size_cut == interval].tip_norm.values

  fig.add_trace(
   go.Bar(
    x= df_2_cut.loc[df_2_cut.order_size_cut == interval].order_size_cut.values,
    y= df_2_cut.loc[df_2_cut.order_size_cut == interval].tip_norm.values,
    name= f"Tipp: {(df_2_cut.loc[df_2_cut.order_size_cut == interval].tip_norm.values[0] * 100).round(2)}%",
    showlegend= True,
    marker= dict(color= color),
    width= 0.8,
    text = text,
    legendgroup= f"{interval}",
    legendgrouptitle_text = f"{interval}",
    hovertemplate = f"<b>{interval}</b><br><br>" +
        "Tipp: %{y:.0%}<br>" +
        "<extra></extra>",
   ),
  )

 else:
  color = px.colors.sequential.Greys[3]

  fig.add_trace(
   go.Bar(
    x= df_2_cut.loc[df_2_cut.order_size_cut == interval].order_size_cut.values,
    y= df_2_cut.loc[df_2_cut.order_size_cut == interval].tip_norm.values,
    name= f"Tipp: {(df_2_cut.loc[df_2_cut.order_size_cut == interval].tip_norm.values[0] * 100).round(2)}%",
    showlegend= True,
    marker= dict(color= color),
    width= 0.8,
    legendgroup= f"{interval}",
    legendgrouptitle_text = f"{interval}",
    hovertemplate = f"<b>{interval}</b><br><br>" +
        "Tipp: %{y:.0%}<br>" +
        "<extra></extra>",
   ),
  )


fig.update_layout(title= dict(text= "<b>Tippwahrscheinlichkeit jeder Bestellungsgröße</b>", font_size= 20, xanchor= "center", yanchor= "top", y=0.935, x=0.475), legend_title_text = "<b>Bestellgröße</b>", height= 800)
fig.update_xaxes(title_text="<b>Bestellgröße in Intervallen</b>", ticks= "outside", col= 1, row= 2, showline= True)
fig.update_yaxes(title_text="<b>Tipp Wshk.</b>", ticks= "outside", showgrid= True, gridcolor= "grey", col= 1, row= 2, showline= True, tickvals= [0, 0.25, 0.5, 0.75])
fig.update_layout(plot_bgcolor= "white", xaxis= dict(linecolor= "black"), yaxis= dict(linecolor= "black"))

fig.show()

- Bestellungen mit ab einer Bestellgröße von >25 haben eine Trinkgeldwahrscheinlichkeit von >= 50%

- Das Maximum der Trinkgeldwahrscheinlichkeit liegt bei  Bestellungen mit einer Bestellgröße von 35 bis 150 Produkten, dabei sei gesagt, dass Intervall größer ist, da in diesem sich nur 0.1% der Bestelldaten befinden

### **P4:** Spielt die Vergangenheit eines Users eine Rolle in Hinblick auf die Trinkgeldwahrscheinlichkeit, oder kommt es nur auf Inhalt und Parameter der aktuellen Bestellung an?

- Gruppieren der Bestellungen nach User und identifizierend der aktuellen Bestellung pro User

- User Trinkgeldwahrscheinlichkeit aus ihrer Bestellhistorie berechnen und mit der Trinkgeldgabe ihrer aktuellen Bestellung vergleichen 

##### Preprocessing

In [ ]:
df_user_hist = df.copy()

In [ ]:
df_user_hist = df_user_hist.sort_values(by= ["user_id", "order_number"], ascending= True)
df_user = df_user_hist.groupby("user_id", as_index= False).agg(orders = ("order_id", "unique"), reordered= ("reordered", "mean"), tip= ("tip", "mean"), order_number= ("order_number", "unique"))
df_user.head()

In [ ]:
last_orders = []

for ind, order in enumerate(df_user.orders):
 last_orders.append(df_user.orders[ind][-1])

df_user_hist.drop(df_user_hist.loc[df_user_hist.order_id.isin(last_orders)].index, inplace= True)

In [ ]:
df_userh = df_user_hist.groupby("user_id", as_index= False).agg(orders = ("order_id", "nunique"), reordered= ("reordered", "mean"), tip= ("tip", "mean"))
df_userh.head()

In [ ]:
df_last_order = df.loc[df.order_id.isin(last_orders)].groupby("user_id").agg(tip_last_order= ("tip", "max"), last_order_id = ("order_id", "max")).reset_index()
df_last_order.head()

In [ ]:
df_userh = df_userh.merge(df_last_order, how= "right", on= "user_id")
df_userh.head()

##### Analyse

**Bestellungen bei welcher der user eine tip_wshk von >= 0.5 hat verglichen mit seiner letzten/aktuellen Bestellung**


Unter der Annahme, dass wenn die tip_wshk eines users in der historie >= 0.5 ist, tendiert er eher dazu zu tippen bei seiner letzten/aktuellen Bestellung

**trifft zu:**

In [ ]:
df_userh.loc[(df_userh.tip >= 0.5) & (df_userh.tip_last_order == 1)].head()

In [ ]:
case1_true = 7685 / 37367
case1_true

**trifft nicht zu:**

In [ ]:
df_userh.loc[(df_userh.tip >= 0.5) & (df_userh.tip_last_order == 0)].head() 

In [ ]:
case1_false = 4181 / 37367
case1_false

**Bestellungen bei welcher der user eine tip_wshk von < 0.5 hat verglichen mit seiner letzten/aktuellen Bestellung**

Unter der Annahme, dass wenn die tip_wshk eines users in der historie < 0.5 ist, tendiert er eher dazu nicht zu tippen bei seiner letzten/aktuellen Bestellung

**trifft zu:**

In [ ]:
df_userh.loc[(df_userh.tip < 0.5) & (df_userh.tip_last_order == 0)].head()

In [ ]:
case2_true = 20635 / 37367
case2_true

**trifft nicht zu:**

In [ ]:
df_userh.loc[(df_userh.tip < 0.5) & (df_userh.tip_last_order == 1)].head() 

In [ ]:
case2_false = 4866 / 37367
case2_false

In [ ]:
print(f"Wahrscheinlichkeit Trifft zu: {case1_true + case2_true}, Wahrscheinlichlkeit Trifft nicht zu: {case1_false + case2_false}")

##### Preprocessing und Feature Importance 

- Um zu belegen das die Vergangenheit eines Users die Trinkgeldwahrscheinlichkeit seiner aktuellen Bestellung beeinlfusst wird der Dataframe in zwei Datensätze unterteilt.   

- Im Ersten sind alle Bestellung der Historie eines Users und im Zweiten befinden sich nur die aktuellen Bestellungen eines Users.  

- Das Feature user_tip_ratio wird erstellt, welches die Tippwahrscheinlichkeit jedes Users bis Zeitpunkt seiner aktuellen Bestellung angibt.    

- Auf dem ersten Datensatz wird nun ein Entscheidungsbaum als Classifier trainiert der die Wahrscheinlichkeit, dass eine Bestellung einen Tipp erhält, (tip: 0/1) auf dem zweiten Datensatz (die aktuellen Bestellungen) versucht hervorzusagen. 

In [ ]:
# Funktion für die Print-Ausgabe des Classifiers

def print_evaluation(pipeline_or_model: str, X_train, X_test, y_train, y_test, y_train_pred, y_test_pred):
    accurary_train = accuracy_score(y_train, y_train_pred)
    recall_train = recall_score(y_train, y_train_pred)
    precision_train = precision_score(y_train, y_train_pred)

    accurary_test = accuracy_score(y_test, y_test_pred)
    recall_test = recall_score(y_test, y_test_pred)
    precision_test = precision_score(y_test, y_test_pred)
    
    print(
        f"{pipeline_or_model} Evaluation:\n"
        f"{'':6} {'ACC':>10} | {'REC':>14} | {'PRE':>10} | {'rows':>8} | {'columns':>8}\n"
        f"{'Train':6} {accurary_train:10.5f} | {recall_train:14.5f} | {precision_train:10.5f} | {X_train.shape[0]:8} | {X_train.shape[1]:8}\n"
        f"{'Test':6} {accurary_test:10.5f} | {recall_test:14.5f} | {precision_test:10.5f} | {X_test.shape[0]:8} | {X_test.shape[1]:8}\n"
    )

**Bestimmen der letzten/aktuellen Bestellung eines Users**

In [ ]:
df_userhh = df.copy()

df_userhh1 = df_userhh.sample(frac= 0.1)

df_userhh2 = df_userhh1.sort_values(by= ["user_id", "order_number"], ascending= True)
df_userh1 = df_userhh2.groupby("user_id", as_index= False).agg(orders = ("order_id", "unique"), tip= ("tip", "mean"), order_number= ("order_number", "unique"))

last_orders = []
for ind, order in enumerate(df_userh1.orders):
 last_orders.append(df_userh1.orders[ind][-1])

**Aufteilen des Dataframes in Bestellungen aus der Vergangenheit und aktuelle Bestellungen**

In [ ]:
df1 = df_userhh1.copy()

In [ ]:
df_last_orders = df1.loc[df1.order_id.isin(last_orders)]

In [ ]:
df1.drop(df1.loc[df1.order_id.isin(last_orders)].index, inplace= True)
df_tip = df1.groupby(["user_id"], as_index= False).agg(user_tip_ratio = ("tip", "mean"))
df_tip.head()

**Zusammenstellen des Trainings und Testdatensatzes**

In [ ]:
df1 = df1.merge(df_tip, on= "user_id", how= "right")
df_last_orders = df_last_orders.merge(df_tip, on= "user_id", how= "right")

**Trainingsdatensatz (Bestellungen vor der aktuelle Bestellung des Users):**

In [ ]:
df1

**Testdatensatz (aktuelle Bestellungen eines Users):**

In [ ]:
df_last_orders

**One Hot Encoding der Spalte days_since_prior_order, da NAN-Values**

- müssen geändert werden da die Methode DeecisionTreeClassifier nicht mit NAN-Werten umgehen kann

In [ ]:
df_train = pd.get_dummies(df1, columns= ["county", "days_since_prior_order"], dummy_na= True)

In [ ]:
df_test = pd.get_dummies(df_last_orders, columns= ["county", "days_since_prior_order"], dummy_na= True)

##### Entscheidungsbaum und Feature Importance

- Trainings- und Testdatensplit

In [ ]:
df_train.drop(["product_name", "department", "aisle"], axis= 1, inplace= True)
df_test.drop(["product_name", "department", "aisle"], axis= 1, inplace= True)

In [ ]:
features_Xtrain = df_train.columns.to_list()
features_Xtrain.remove("tip")

In [ ]:
features_Xtest = df_test.columns.to_list()
features_Xtest.remove("tip")

In [ ]:
X_train, y_train = df_train[features_Xtrain], df_train["tip"]

In [ ]:
X_test, y_test = df_test[features_Xtest], df_test.pop("tip")

- Mittels einer 5-fachen Kreuzvaliedierung und Hyperparameteroptimierung wird die Beste Kombination zwischen Baumtiefe (max_depth) und minimaler Anzahl an benötigter Beispiele für einen Split (min_samples_split) gesucht

In [ ]:
gs_crossv = GridSearchCV(
 estimator= DecisionTreeClassifier(),
 param_grid= dict(min_samples_split= [2, 10, 20, 50, 100], max_depth= [1, 3, 6], random_state= [0]),
 scoring = "accuracy",
 n_jobs= -1,
 cv= 5
).fit(X_train, y_train)

print(f"Best parameters: {gs_crossv.best_params_}, with score: {gs_crossv.best_score_}")

In [ ]:
dtc2 = DecisionTreeClassifier(min_samples_split= 2, max_depth= 6, random_state= 0).fit(X_train, y_train)

y_pred_train_best_dtc2 = dtc2.predict(X_train)
y_pred_test_best_dtc2 = dtc2.predict(X_test)

print_evaluation("DTC", X_train, X_test, y_train, y_test, y_pred_train_best_dtc2, y_pred_test_best_dtc2)

**Das Modell schneidet in der Evaluation besser ab, als das Basismodell ohne Feature Engineering in P3**

In [ ]:
print(f"Feature Importance: {dict(zip(X_train.columns, dtc2.feature_importances_))}")

In [ ]:
feature_importances_p4 = pd.DataFrame.from_dict(dict(zip(X_train.columns, dtc2.feature_importances_)), orient= "index", columns= ["feature_importance"])

feature_importances_sorted = feature_importances_p4.sort_values(by= "feature_importance", ascending= True).tail(9)
df_feature_imp_p4 = feature_importances_sorted.reset_index().rename(columns= {"index": "feature"})

In [ ]:
df_feature_imp_p4

In [ ]:
# P3 Plot: Feature Importance

fig = go.Figure()


for feature in df_feature_imp_p4.feature.unique():

 if df_feature_imp_p4.loc[df_feature_imp_p4.feature == feature].feature_importance.values >= 0.1:
  color= px.colors.sequential.Purpor[5]

  fig.add_trace(
   go.Bar(
    x= df_feature_imp_p4.loc[df_feature_imp_p4.feature == feature].feature_importance,
    y= df_feature_imp_p4.loc[df_feature_imp_p4.feature == feature].feature,
    marker= dict(color= color),
    orientation = "h",
    text= df_feature_imp_p4.loc[df_feature_imp_p4.feature == feature].feature_importance.round(2),
    showlegend= False,
    width= 0.8,
    hovertemplate= f"<b>{feature}</b><br><br>" +
                   "feature_importance: %{x}<br>" +
                   "<extra></extra>"
    )
  )
 else:
  color= px.colors.sequential.Greys[3]

  fig.add_trace(
   go.Bar(
    x= df_feature_imp_p4.loc[df_feature_imp_p4.feature == feature].feature_importance,
    y= df_feature_imp_p4.loc[df_feature_imp_p4.feature == feature].feature,
    marker= dict(color= color),
    orientation = "h",
    text= df_feature_imp_p4.loc[df_feature_imp_p4.feature == feature].feature_importance.round(3),
    showlegend= False,
    width= 0.8,
    hovertemplate= f"<b>{feature}</b><br><br>" +
                   "feature_importance: %{x}<br>" +
                   "<extra></extra>"
    )
  )

fig.update_layout(title= dict(text= "<b>Feature Importance der Attribute</b>", font_size= 20, xanchor= "center", yanchor= "top", y=0.935, x=0.475), height= 600)
fig.update_xaxes(title_text="<b>Importance</b>", ticks= "outside")
fig.update_xaxes(showline= True)
fig.update_yaxes(title_text="<b>Feature</b>", ticks= "outside", showgrid= True, gridcolor= "grey", showline= True)
fig.update_layout(plot_bgcolor= "white", xaxis= dict(linecolor= "black"), yaxis= dict(linecolor= "black"))
fig.show()

- feature_importance von user_tip_ratio mit 0.9379 am höchsten
 
- **damit ist die Tip-Historie eines Users auschschlaggebend für die Tippwahrscheinlichkeit bei seiner aktuellen Bestellung** 

## 3. Zusätzliche Fragestellungen:

#### Machine Learning Modell:

##### Preprocessing und Feature Engineering:

In [ ]:
df_g = df.copy()

**Nan-values, welche in der Spalte days_since_prior_orders auftreten wenn es die erste Bestellung eines Kunden ist, für die Methode Decision Tree Classifier mit 0 auffüllen**

In [ ]:
df = df_g.reset_index(drop= True)

**Sampling des Datensatzes um Rechenzeit zu sparen**

In [ ]:
df_sample = df.sample(frac= 0.1)

**Erzeugen des Features user_tip_raito**

In [ ]:
df_tip = df_sample.groupby(["user_id", "order_id", "order_number"], as_index= False).agg(tip = ("tip", "mean")).sort_values(by= ["user_id", "order_number"], ascending= True)
df_tip["tip_cumsum"] = df_tip.groupby("user_id", as_index= False)["tip"].cumsum()
df_tip["user_tip_ratio"] = df_tip.tip_cumsum / df_tip.order_number

df_tip_merge = df_tip[["user_id", "order_id", "order_number", "user_tip_ratio"]]
df_sample = df_sample.merge(df_tip_merge, on = ["user_id", "order_id", "order_number"], how= "right")

In [ ]:
df_classifier = df_sample.copy()

**One-Hot-Encoding der Spalte County**

In [ ]:
df_classifier_dtc = pd.get_dummies(df_classifier, columns= ["county", "days_since_prior_order"], dummy_na= True)

In [ ]:
df_classifier_ghbc = pd.get_dummies(df_classifier, columns= ["county"])

In [ ]:
X_dtc = df_classifier_dtc.drop(["product_name", "department", "aisle", "tip"], axis= 1)
y_dtc = df_classifier_dtc["tip"]

X_ghbc = df_classifier_ghbc.drop(["product_name", "department", "aisle", "tip"], axis= 1)
y_ghbc = df_classifier_dtc["tip"]

In [ ]:
X_train_d, X_test_d, y_train_d, y_test_d = train_test_split(X_dtc, y_dtc, test_size= 0.2, random_state=0)

In [ ]:
X_train_h, X_test_h, y_train_h, y_test_h = train_test_split(X_ghbc, y_ghbc, test_size= 0.2, random_state=0)

##### **HGBC**

In [ ]:
gs_crss_hgbc = GridSearchCV(
 estimator= HistGradientBoostingClassifier(),
 param_grid= dict(max_iter= [100], learning_rate= [0.001, 0.01, 0.1], max_depth= [1,3,5,9], random_state= [0]),
 scoring= "accuracy",
 n_jobs= -1
).fit(X_train_h, y_train_h)

print(f"Best param: {gs_crss_hgbc.best_params_} with acc: {gs_crss_hgbc.best_score_}")

In [ ]:
best_hgbc = HistGradientBoostingClassifier(learning_rate= 0.1, max_depth= 9, max_iter= 100, random_state= 0).fit(X_train_h, y_train_h)

y_best_train_pred = best_hgbc.predict(X_train_h)
y_best_test_pred = best_hgbc.predict(X_test_h)

print_evaluation("HGBC", X_train_h, X_test_h, y_train_h, y_test_h, y_best_train_pred, y_best_test_pred)

##### **DTC**

In [ ]:
gs_crss_dtc = GridSearchCV(
 estimator= DecisionTreeClassifier(),
 param_grid= dict(min_samples_split= [2, 10, 20, 50, 100], max_depth= [1, 3, 5, 9], random_state= [0]),
 scoring= "accuracy",
 n_jobs= -1
).fit(X_train_d, y_train_d)

print(f"Best param: {gs_crss_dtc.best_params_} with acc: {gs_crss_dtc.best_score_}")

In [ ]:
best_dtc = DecisionTreeClassifier(max_depth= 9, min_samples_split= 2, random_state= 0).fit(X_train_d, y_train_d)

y_best_train_pred = best_dtc.predict(X_train_d)
y_best_test_pred = best_dtc.predict(X_test_d)

print_evaluation("DTC", X_train_d, X_test_d, y_train_d, y_test_d, y_best_train_pred, y_best_test_pred)